Zero-size WAL files are not removed on exit

Hello there,

QL is great, and I can use it without any GCC dependencies. I noticed that I have several .dot files being created in the same folder where I run a program that uses QL library (see below), but they are not removed after the program exists, and I do call `defer _DB.close()` in my main. I think these files might be related to QL locking mechanism.
```
.34a8b4c1ad933745146fdbfef3073706ee571625
.8750451be59059bc01b258e5315f2e7aa75d4350
.f630cc97640448d8f823924a7df3ca00e4d9e53b
```

Is it the intended behavior, or do I need to do anything in my code to clean things up?

I ran the tests before making any changes, and it produces some errors - I assume they are all known issues (at least on Windows)?

```
$ go test -tags purego
--- FAIL: TestMemStorage (0.76s)
        storage_test.go:485: FAIL: test # 300
                -- 300
                BEGIN TRANSACTION;
                        CREATE TABLE t(c blob);
                        INSERT INTO t VALUES (blob(`
                0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
                `));
                COMMIT;
                SELECT * FROM t;

                ---- g
                "c"
                [[13 10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 10                0 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 13 1                0]]
                ---- e
                "c"
                [[10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 1                01 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 10]]
                ----
--- FAIL: TestFileStorage (8.31s)
        storage_test.go:485: FAIL: test # 300
                -- 300
                BEGIN TRANSACTION;
                        CREATE TABLE t(c blob);
                        INSERT INTO t VALUES (blob(`
                0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
                `));
                COMMIT;
                SELECT * FROM t;

                ---- g
                "c"
                [[13 10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 10                0 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 13 1                0]]
                ---- e
                "c"
                [[10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 1                01 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 10]]
                ----
--- FAIL: TestOSFileStorage (8.25s)
        storage_test.go:485: FAIL: test # 300
                -- 300
                BEGIN TRANSACTION;
                        CREATE TABLE t(c blob);
                        INSERT INTO t VALUES (blob(`
                0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
                `));
                COMMIT;
                SELECT * FROM t;

                ---- g
                "c"
                [[13 10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 13 10]]
                ---- e
                "c"
                [[10 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 48 49 50 51 52 53 54 55 56 57 97 98 99 100 101 102 10]]
                ----
--- FAIL: TestIssue28 (0.01s)
        all_test.go:2018: parse file://C:\Users\KRYUKO~1\AppData\Local\Temp\ql-test-657924690\ql.db: invalid character "\\" in host name
--- FAIL: TestIssue56 (0.01s)
        all_test.go:2241: parse file://C:\Users\KRYUKO~1\AppData\Local\Temp\ql-test-079864660\test.db: invalid character "\\" in host name
--- FAIL: TestIssue66FileDriver (0.01s)
        all_test.go:2537: parse file://C:\Users\KRYUKO~1\AppData\Local\Temp\ql-test-766154662\TestIssue66MemDriver: invalid character "\\" in host name
--- FAIL: TestIssue73 (0.01s)
        all_test.go:2621: Call to db.Begin()
        all_test.go:2623: db.Begin:  parse file://C:\Users\KRYUKO~1\AppData\Local\Temp\ql-test-812338893\test.db: invalid character "\\" in host name
--- FAIL: TestHTTP (0.00s)
        httpfs_test.go:123: file does not exist
FAIL
exit status 1
FAIL    github.com/vkryukov/ql  19.397s
```

--- a/CONTRIBUTORS
+++ b/CONTRIBUTORS
@@ -10,4 +10,5 @@ Dan Kortschak <dan.kortschak@adelaide.edu.au>
 Geofrey Ernest <geofreyernest@live.com>
 Jan Mercl <0xjnml@gmail.com>
 OpenNota <opennota@gmail.com>
+Victor Kryukov <victor.kryukov@gmail.com>
 Viktor Kojouharov <vkojouharov@gmail.com>

--- a/file.go
+++ b/file.go
@@ -101,6 +101,8 @@ func OpenFile(name string, opt *Options) (db *DB, err error) {
 		}
 	}
 
+	fi.removeEmptyWAL = opt.RemoveEmptyWAL
+
 	return newDB(fi)
 }
 
@@ -129,16 +131,22 @@ func OpenFile(name string, opt *Options) (db *DB, err error) {
 //
 // Headroom
 //
-// Headroom selecst the minimum size a WAL file will have. The "extra"
+// Headroom selects the minimum size a WAL file will have. The "extra"
 // allocated file space serves as a headroom. Commits that fit into the
 // headroom should not fail due to 'not enough space on the volume' errors. The
 // headroom parameter is first rounded-up to a non negative multiple of the
 // size of the lldb.Allocator atom.
+//
+// RemoveEmptyWAL
+//
+// RemoveEmptyWAL controls whether empty WAL files should be deleted on
+// clean exit.
 type Options struct {
-	CanCreate bool
-	OSFile    lldb.OSFile
-	TempFile  func(dir, prefix string) (f lldb.OSFile, err error)
-	Headroom  int64
+	CanCreate      bool
+	OSFile         lldb.OSFile
+	TempFile       func(dir, prefix string) (f lldb.OSFile, err error)
+	Headroom       int64
+	RemoveEmptyWAL bool
 }
 
 type fileBTreeIterator struct {
@@ -383,16 +391,17 @@ func (t *fileTemp) Set(k, v []interface{}) (err error) {
 }
 
 type file struct {
-	a        *lldb.Allocator
-	codec    *gobCoder
-	f        lldb.Filer
-	f0       lldb.OSFile
-	id       int64
-	lck      io.Closer
-	mu       sync.Mutex
-	name     string
-	tempFile func(dir, prefix string) (f lldb.OSFile, err error)
-	wal      *os.File
+	a              *lldb.Allocator
+	codec          *gobCoder
+	f              lldb.Filer
+	f0             lldb.OSFile
+	id             int64
+	lck            io.Closer
+	mu             sync.Mutex
+	name           string
+	tempFile       func(dir, prefix string) (f lldb.OSFile, err error)
+	wal            *os.File
+	removeEmptyWAL bool // Whether empty WAL files should be removed on close
 }
 
 func newFileFromOSFile(f lldb.OSFile, headroom int64) (fi *file, err error) {
@@ -596,12 +605,22 @@ func (s *file) Close() (err error) {
 
 	es := s.f0.Sync()
 	ef := s.f0.Close()
-	var ew error
+	var ew, estat, eremove error
 	if s.wal != nil {
+		remove := false
+		wn := s.wal.Name()
+		if s.removeEmptyWAL {
+			var stat os.FileInfo
+			stat, estat = s.wal.Stat()
+			remove = stat.Size() == 0
+		}
 		ew = s.wal.Close()
+		if remove {
+			eremove = os.Remove(wn)
+		}
 	}
 	el := s.lck.Close()
-	return errSet(&err, es, ef, ew, el)
+	return errSet(&err, es, ef, ew, el, estat, eremove)
 }
 
 func (s *file) Name() string { return s.name }

--- a/file_test.go
+++ b/file_test.go
@@ -1,0 +1,40 @@
+// Copyright (c) 2014 ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ql
+
+import (
+	"os"
+	"testing"
+)
+
+func fileExists(p string) bool {
+	_, err := os.Stat(p)
+	return err == nil
+}
+
+func TestWALRemoval(t *testing.T) {
+	const tempDBName = "./_test_was_removal.db"
+	wName := walName(tempDBName)
+	defer os.Remove(tempDBName)
+	defer os.Remove(wName)
+
+	db, err := OpenFile(tempDBName, &Options{CanCreate: true})
+	if err != nil {
+		t.Fatalf("Cannot open db %s: %s\n", tempDBName, err)
+	}
+	db.Close()
+	if !fileExists(wName) {
+		t.Fatalf("Expect WAL file %s to exist but it doesn't", wName)
+	}
+
+	db, err = OpenFile(tempDBName, &Options{CanCreate: true, RemoveEmptyWAL: true})
+	if err != nil {
+		t.Fatalf("Cannot open db %s: %s\n", tempDBName, err)
+	}
+	db.Close()
+	if fileExists(wName) {
+		t.Fatalf("Expect WAL file %s to be removed but it still exists", wName)
+	}
+}

